### PR TITLE
Improvements to code contracts [depends-on: #2646, #2676, blocks: #2712]

### PR DESCRIPTION
**Pull request by qaphla**
This provides a few changes which together serve to improve the havocking of variables when dealing with function calls (as well as a patch [02eb459] which doesn't really seem to fit anywhere else).

63b2af3 fixes a soundness issue with contract application by havocking the variables which could be touched by the function call being abstracted out.
2ec8f88 resolves an issue wherein function contracts were checked before initializing the dead and deallocated objects, and so any check that a pointer was not dead or deallocated would fail.
c3fcb36 switches contracts from using the local_may_alias analysis to the goto_rw analysis for finding the write sets of functions or loops.
a9354ac  improves the documentation of the contracts feature.

This builds on https://github.com/diffblue/cbmc/pull/2646 because the changes in c3fcb36 benefit greatly from a less conservative approximation in goto_rw.

Only the commits mentioned by name in this are new. The others either come from https://github.com/diffblue/cbmc/pull/2646 or https://github.com/diffblue/cbmc/pull/2676.

### CODE REVIEW

**Review by allredj — COMMENTED** [review]
This PR failed Diffblue compatibility checks (cbmc commit: 522c765291533358fa274cd272881fa2fbaa21c4).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/80904637
Status will be re-evaluated on next push.
Please contact @peterschrammel, @thk123, or @allredj for support.

Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).
##

File: .gitignore [hunk 1]
```diff
@@ -65,6 +67,8 @@ jbmc/regression/**/tests-symex-driven-loading.log
 
 # files stored by editors
 *~
+.*.swp
+.*.swo
```
Comment (kroening) [inline .gitignore:72]: Is this not redundant with the above?

File: src/analyses/dependence_graph.cpp [hunk 2]
```diff
@@ -184,6 +186,8 @@ void dep_graph_domaint::data_dependencies(
             data_deps.insert(w_range.first);
             found=true;
           }
+        }
+      }
     }
```
Comment (kroening) [inline src/analyses/dependence_graph.cpp:191]: The change to this file appears unrelated.

File: src/analyses/goto_rw.cpp [hunk 3]
```diff
@@ -414,16 +412,19 @@ void rw_range_sett::get_objects_typecast(
 
 void rw_range_sett::get_objects_address_of(const exprt &object)
 {
-  if(object.id() == ID_string_constant ||
+  if(object.id() == ID_symbol ||
+     object.id() == ID_string_constant ||
      object.id() == ID_label ||
```
Comment (kroening) [inline src/analyses/goto_rw.cpp:417]: The ID_symbol case overrides the case below.

File: src/ansi-c/c_typecheck_expr.cpp [hunk 4]
```diff
@@ -2141,7 +2141,7 @@ exprt c_typecheck_baset::do_special_functions(
       throw 0;
     }
 
-    exprt dynamic_object_expr=exprt(ID_dynamic_object, expr.type());
+    exprt dynamic_object_expr=exprt(ID_is_dynamic_object, expr.type());
     dynamic_object_expr.operands()=expr.arguments();
```
Comment (kroening) [inline src/ansi-c/c_typecheck_expr.cpp:2145]: Use is_dynamic_object_exprt, which you are adding.

**Review by kroening — CHANGES_REQUESTED** [review]

**Comment by kroening** [comment]
Given that new syntax is proposed here, did you consider ACSL?
https://github.com/acsl-language/acsl

**Review by martin-cs — COMMENTED** [review]
Although I appreciate it is a pain, could we get the supporting PRs merged first and then look at this?  I'm also very much with @kroening on the idea of annotation languages : the world may need another annotation language for C but it won't be one that is a partial subset of the features of an existing one.  Interoperability is good and being able to compare wih / share benchmarks with tools like Frama-C would be great.

**Comment by TGWDB** [comment]
Closing due to age (no further comment on PR content), please reopen with rebase on develop if you intent to continue this work.

**Pull request closed without merging**